hugolib/page.go

```diff
@@ -800,16 +800,16 @@ func (p *Page) Render(layout ...string) template.HTML {
 	return tpl.ExecuteTemplateToHTML(p, l...)
 }
 
-func (p *Page) guessMarkupType() string {
+func (p *Page) determineMarkupType() string {
 	// First try the explicitly set markup from the frontmatter
 	if p.Markup != "" {
-		format := helpers.GuessType(p.Markup)
-		if format != "unknown" {
-			return format
+		p.Markup = helpers.GuessType(p.Markup)
+		if p.Markup == "unknown" {
+			p.Markup = helpers.GuessType(p.Source.Ext())
 		}
 	}
 
-	return helpers.GuessType(p.Source.Ext())
+	return p.Markup
```

Comment:
This line is wrong.  If `p.Markup == ""`, you return `""` instead of `"unknown"`, which [shortcodes are not prepared for](https://github.com/spf13/hugo/blob/a89035bdaaa8bb1525a74d82e068ef80bfa28aed/hugolib/shortcode.go#L254).  Follow the original logic:

```go
	// First try the explicitly set markup from the frontmatter
	if p.Markup != "" {
		p.Markup = helpers.GuessType(p.Markup)
		if p.Markup != "unknown" {
			return p.Markup
		}
	}

	p.Markup = helpers.GuessType(p.Source.Ext())
	return p.Markup
```